File: leetcode/000543.diameter-of-binary-tree.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val
		this.left = left === undefined ? null : left
		this.right = right === undefined ? null : right
	}
}

function diameterOfBinaryTree(root: TreeNode | null): number {
	let maxDiameter = 0

	function dfs(root: TreeNode | null) {
		if (!root) {
			return -1
		}

		const left = dfs(root.left)
		const right = dfs(root.right)

		maxDiameter = Math.max(maxDiameter, left + right + 1 + 1) // Height of left subtree + height of right subtree + 1 to cancel -1 height for null node + 1 for current node

		return 1 + Math.max(left, right)
	}

	dfs(root)

	return maxDiameter
}
// Time complexity: O(n)
// Space complexity: O(n)

File: leetcode/python/000543.diameter-of-binary-tree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxDiameter = 0

        def dfs(node: Optional[TreeNode]) -> int:
            if node is None:
                return 0

            leftH = dfs(node.left)
            rightH = dfs(node.right)
            diameter = leftH + rightH

            nonlocal maxDiameter
            maxDiameter = max(maxDiameter, diameter)  # updates maxDiameter

            return max(leftH, rightH) + 1  # returns the higher height, + 1 for self

        dfs(root)
        return maxDiameter
